src/fhir_search/uri_parser.clj

```diff
@@ -0,0 +1,140 @@
+
+(defn values [query]
+  (->>
+   (str/split query #"&")
+   (reduce (fn [result element]
+             (let [group (str/split element #"=")
+                   first-group (first group)
+                   second-group (second group)]
+               (if (re-find #"," second-group)
+                 (conj result (->>
+                               (reduce (fn [value item]
+                                         (conj value {:modifiers (when (second (str/split first-group #":"))
+                                                                   (keyword "fhir.search.modifier" (second (str/split first-group #":"))))
+                                                      :prefix (when (prefix-find prefixes item) (keyword "fhir.search.prefix" (prefix-find prefixes item)))
+                                                      :value (if (prefix-find prefixes item)
+                                                               (str/replace item (re-pattern (prefix-find prefixes item)) "")
+                                                               item)}))
+                                       [] (str/split second-group #","))
+                               (assoc {:join :fhir.search.join/or
+                                       :name (first (str/split first-group #":"))} :values)))
+
+                 (conj result {:name (first (str/split first-group #":"))
+                               :modifiers (when (second (str/split first-group #":"))
+                                            (keyword "fhir.search.modifier" (second (str/split first-group #":"))))
+                               :prefix (when (prefix-find prefixes second-group) (keyword "fhir.search.prefix" (prefix-find prefixes second-group)))
+                               :value (if (prefix-find prefixes second-group)
+                                        (str/replace second-group (re-pattern (prefix-find prefixes second-group)) "")
+                                        second-group)})))) [])))
+
+(defn path [string]
+  (let [path (remove empty? (-> (.getPath (URI. string)) (str/split #"/")))
+        structure {:type (last path)}]
+    (cond-> structure
+      (= 3 (count path)) (assoc :compartment {:type (first path) :id (second path)})
+      (= 2 (count path)) (assoc :component {:id (first path)}))))
+
+(defn params [string]
+  (let [query (.getQuery (URI. string))]
+    (when query
+      (hash-map :params {:join :fhir.search.join/and
+                         :values (values query)}))))
+
+(defn clean [m]
+  (postwalk (fn [v]
+              (cond
+                (instance? clojure.lang.MapEntry v) (when-not (nil? (val v)) v)
+                (map? v) (when-let [entries (seq (remove #(-> % second nil?) v))] (into {} entries))
+                (vector? v) (when-let [coll (seq (remove nil? v))] (into [] coll))
+                (seq? v) (remove nil? v)
+                :else v)) m))
+
+(defn uri-parser [url] (postwalk clean (into (path url) (params url))))
```

Comment:
The name parser, is not fair here, parser suggests the responsibility for parsing, not the action of parsing itself. But we are actually parsing already. We can name is just `parse`.